### DISCUSSION TITLE
Not exactly on the beat, how to finetune it?

### DISCUSSION THREAD

**Discussion by pujiii** [sandarshsridhar/wiz-spotify-connected#32] [General]
The light is not going exactly on the beat and it's kind of bothering me, where could I make an adjustment to fine-tune this, so to shift everything a few milliseconds forwards/backwards?

**Comment by sandarshsridhar** [comment]
@pujiii, I believe the lights are off-beat because your bulbs could be from a batch where the connectivity is laggy. I have personally had this issue with 2 of my bulbs. Wiz bulbs are made by multiple manufacturers. So, there is no consistency in the behavior. 

You could try running the app in `[calm](https://github.com/sandarshsridhar/wiz-spotify-connected#calm)` mode. It should reduce the delay issue since the lights are changed at 1/2 the speed. 

Answering your question about fine-tuning, no explicit delay is specified in the code for changing the lights. The app queues the commands to the bulbs as the beat changes. 

Although technically, there is a delay attribute you could play with (I am not sure if that would help): [SPOTIFY_POLLING_DELAY_MS](https://github.com/sandarshsridhar/wiz-spotify-connected/blob/main/sample-env.txt#L11). This delay is used to wait before making `getCurrentlyPlaying` API call to Spotify. The beats in a song are segmented into sections. And these sections don't change that often. So, the delay is used to make the app efficient by preserving the Spotify API rate limit and avoiding unnecessary frequent I/O calls. The default value is 1000 milliseconds (which means, the app requests Spotify to get the current playback state at 1-second intervals).

**Reply by sandarshsridhar** [reply]
@pujiii, I misspoke earlier, there is a way to fine-tune the beats-to-lights but it's not via configuration. You will have to change the code manually. If you want to try that, here it is https://github.com/sandarshsridhar/wiz-spotify-connected/blob/main/src/utils/dance-engine.ts#L70. Just subtract a few milliseconds depending on what works best for you so that the delay in the bulb's response gets nullified.

**Reply by pujiii** [reply]
@sandarshsridhar I tried finetuning it via that, but this caused the BPM to be adjusted. It caused the beats per minute to change, so the light would blink faster or slower, instead of shifting the whole sequence.

I dug a bit further into your code and it seems that you are just waiting the time that there should be between beats. I think this should almost always cause this issue by design, because when you start with the first beat just determines when the rest will go, so if this loops starts not when a beat should start, the whole song will be off beat, and this amount will vary per song.

A way to solve this would have to completely change how this is handled, by determining at which second each beat should be played of the song, for example: 1.2s, 1.6s, 2.0s, etc., and then using the progress which we receive from the spotify API call to know when we should light up the lights. After the first beat the current method could be used. In order to get this on beat, you would also have to take into account the time it took to make the API call, because song beats are very sensitive to minor deviations.

**Reply by sandarshsridhar** [reply]
@pujiii, I am not sure why adding a slight delay to `delayMs` would cause such a major issue because, in calm mode, the app changes the lights at half the speed with no desynchronization. I haven't tried adding the noise myself to verify that. So, I will believe you. 

I did think about the app starting at a random starting point while developing the app. After I tested the app several hundred times, I didn't see any noticeable difference in the synchronization. Granted, it could be because of my bulbs (they have close-to-nothing latency, and I have 24 of them). Hence, I didn't modify it any further. As you rightly pointed out, this varies largely on the tempo of the song. 

Moreover, if I do implement this, the issue will not completely go away if I change any kind of playback midway through a song. The only way the sync could be 100% correct is by referencing Spotify API's segments instead of sections as they are much smaller compartments of a song. And, going through those many segments every single iteration to pick the right segment would be very compute-intensive with very little gain.

And considering the latency of Spotify API is going to be too much as we can never correctly predict what it could be. Of course, the HTTP client can give me that information after every call. But the amount of complexity this would add to the app is not worth it as I have seen the API to be very fast. 

Nonetheless, I always welcome PRs for improvements. So, please feel free to create one :)

Cheers!
